Tests/Creation/Frame/Content/String/ID3ChapterTypeFrameCreatorTest.swift

```diff
@@ -0,0 +1,70 @@
+//
+//  ID3ChapterTypeFrameCreatorTest.swift
+//  ID3TagEditor
+//
+//  Created by Nolaine Crusher on 4/6/20.
+//  Copyright © 2020 Fabrizio Duroni. All rights reserved.
+//
+
+import XCTest
+@testable import ID3TagEditor
+
+class ID3ChapterTypeFrameCreatorTest: XCTestCase {
+    func testCreatorFrameChapterType() {
+        let id3ChapterTypeFrameCreator = ID3ChapterTypeFrameCreator(
+            frameContentSizeCalculator: ID3FrameContentSizeCalculator(
+                uInt32ToByteArrayAdapter: UInt32ToByteArrayAdapterUsingUnsafePointer(),
+                synchsafeEncoder: SynchsafeIntegerEncoder()
+            ),
+            frameFlagsCreator: ID3FrameFlagsCreator(),
+            stringToBytesAdapter: ID3UTF16StringToByteAdapter(
+                paddingAdder: PaddingAdderToEndOfContentUsingNullChar(),
+                frameConfiguration: ID3FrameConfiguration()
+            ), frameCreatorsChain: ID3FrameCreatorsChain()
+        )
+        
+        let frameBytes = id3ChapterTypeFrameCreator.createFrame(
+            frameIdentifier: [0x22],
+            version: .version3,
+            elementID: "ch01",
+            startTime: 5000,
+            endTime: 10000,
+            startingByteOffset: 0,
+            endingByteOffset: 0,
+            embeddedSubframes: [.Title : ID3FrameWithStringContent(content:  "Chapter 01")]
+        )
+        
+//        let id3FrameFromStringContentCreator = ID3FrameFromStringContentCreator(
```

Comment:
I used this commented out code to get the bytes I used for the test. Again, not sure thats the right way to handle it.

Reply:
I’m not sure I follow. Start by reading in a valid file, and printing the entire frame’s bytes before you try to process them. From there (and with the help of the specification) you figure out how those bytes mean what they mean, and you build your parser accordingly.

Reply:
I don't have any parsers in place yet for this frame, I was just trying to see if the `frameCreator` was working first before I moved on to that. So I'm not sure it's possible yet to print the frames bytes, but I will work on that.

Reply:
Yes I saw that. It’s just easier (and less error‐prone) to write the parser first and then the encoder. Same as it’s easier to learn to read than to learn to write.

Reply:
...lol well, that was silly of me. I've been doing it the other way around this whole time (probably because the frame creator side of things seemed less daunting than the morass of all the parser stuff.)

Reply:
Question:

In `FrameParser` are the header and encoding considered indices 0 and 1 when locating items by index? So, instead of:

```swift
    internal func extractChapterElements(
        from frameData: inout Data.SubSequence,
        encoding: ID3StringEncoding) -> (elementID: String, startTime: Int, endTime: Int, startByteOffset: Int, endByteOffset: Int, embeddedSubframes: [ID3Frame]?
        ) {
            guard let elementID = frameData.extractPrefixAsStringUntilNullTermination(encoding) else { return "chXX" }
            guard let startTime = frameData.index(1, offsetBy: 0)
            guard let endTime = frameData.index(2, offsetBy: 0)
            guard let startByteOffset = frameData.index(3, offsetBy: 0)
            guard let endByteOffset = frameData.index(4, offsetBy: 0)
            guard let embeddedSubframes = frameData.index(5, offsetBy: 0)

    }
```

`startTime` would actually be at index 3?

Reply:
No. `Data.SubSequence` has arbitrary indices. (`1` might be out of bounds before the relevant subsection of the data even starts.)

You’ll need to use `frameData.index(frameData.startIndex, offsetBy: x, limitedBy: frameData.endIndex)`, where x is `0`, `1`, `2`, `3`, etc.

The returned index will be optional. That is because you cannot know for sure that the subsection of the data will actually be as long as the specification says (such as if the file being read is corrupt).

Reply:
Okay, do I need to take the header and encoding into account when calculating x?

Reply:
No. Also each unit of offset is one byte, not one unit of meaning. `Data` is more‐or‐less an array of bytes.

Reply:
Hm okay. That's going to make it challenging to parse, because there's no way of knowing how many bytes the values of `startTime`, `endTime`, `startByteOffset` and `endByteOffset` will be.

The spec calls for `startTime` and `endTime` to be in milliseconds. That is milliseconds from the start of the audio file, so if we're talking about, say, a 20-hour long audiobook, it could be anything from 0 to 72,000,000.

`startByteOffset` and `endByteOffset` are described as such:

> The Start offset is a zero-based count of bytes from the beginning of the file to the first byte of the first audio frame in the chapter. If these bytes are all set to 0xFF then the value should be ignored and the start time value should be utilized.
> 
> The End offset is a zero-based count of bytes from the beginning of the file to the first byte of the audio frame following the end of the chapter. If these bytes are all set to 0xFF then the value should be ignored and the end time value should be utilized. 

Realistically, they'll probably never be used? Most people will ignore them and use the time stamps instead, but they are in there nonetheless.

Reply:
> That's going to make it challenging to parse, because there's no way of knowing how many bytes the values of startTime, endTime, startByteOffset and endByteOffset will be.

The specification says that each is always exactly 4‐bytes long and represents an integer (`Int64`). The fact that two of the integers happen to describe the byte offsets of other things is irrelevant.

Reply:
Okay, I don't think I understand how the interger-to-bytes conversion works, then. Because what I was doing yesterday with these parameters was converting the STRING of the Integer to bytes.

So, to use my example in the test:

```
            startTime: 5000,
            endTime: 10000,
```

were being converted to:

```
                0x35, 0x30, 0x30, 0x30, // 5000
                0x31, 0x30, 0x30, 0x30, 0x30, // 10000
```

So clearly I'm doing that wrong?

Reply:
Am I understanding the way your `Data.Subsequence` extension is working in assuming that using `extractPrefixAsStringUntilNullTermination` for the `elementID` means that the offset for `startTime` will be 0 and then `endTime` would be 4, etc?

Reply:
To learn more about how integers are stored in memory, look up terms like “fixed width integer” and “binary integer”. I cannot tell from scanning the specification whether it is supposed to be big or little endian, or whether or not it is signed. (Or even if it is floating point instead of an integer?)

But I’m reasonably certain it isn’t an ASCII string, because the highest that could express in four bytes would be `"9999"`, which in milliseconds is only equivalent to 10 seconds. On the other hand, a 32‐bit (4‐byte) unsigned binary integer can hold a value up to (2 ↑ 32) − 1, which is 4 294 967 295 ms or almost 50 days.

Reply:
> Am I understanding the way your Data.Subsequence extension is working in assuming that using extractPrefixAsStringUntilNullTermination for the elementID means that the offset for startTime will be 0 and then endTime would be 4, etc?

Yes. After the element ID is removed out of the front, the remaining data will start with the `startTime`, and after its 4 bytes will be `endTime`, and so on.

Reply:
from looking over what little I could understand from `TagLib` (I was trying to make sense of it to see if I could glean any guidance) I can at least say that they're probably unsigned (I remember the word sticking out at me from a bunch of code I couldn't make sense of [here](taglib/mpeg/id3v2/frames/chapterframe.cpp).) As for big or little endian, I have no idea.

Since most of this is being converted to [UInt8] I'm going to need to convert [UInt32] to [UInt8]. *adds to mental checklist*

Reply:
> I'm going to need to convert [UInt32] to [UInt8].

Rather `UInt32` ↔ `[UInt8]`.